### DISCUSSION TITLE
watchdog: Limitation with the current watchdog API for Nordic devices

### DISCUSSION THREAD

**Discussion by Olivier-ProGlove** [zephyrproject-rtos/zephyr#37740] [General]
**Context:** At boot time my device can take a lot of time to initialize itself - mainly due to flash accesses by NVS and FCB modules.

While I initially setup the watchdog after all these initialization steps, it has recently been decided to move watchdog initialization at the early stage.
Unfortunately the watchdog window is too small for the boot phase.
So I was thinking to setup a first watchdog at boot time with a large time window, then to disable this boot phase watchdog with `wdt_disable()` and setup a new watchdog with a smaller time window.
Unfortunately Nordic watchdog peripheral does not support to be disabled: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/watchdog/wdt_nrfx.c#L65

I cannot keep the small window for the boot phase as I would need to feed the watchdog in NVS/FCB modules. And I do not want to change upstream modules to not have to maintain this delta.
The watchdog window I have in mind is too big for the running phase.

I have few solutions in mind:
* _Risky version:_ Do not initialize the watchdog during the boot phase
* _Hacky version:_ Directly access Nordic Watchdog registers in my Zephyr application to change the Reload value (http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52840.ps/wdt.html?cp=2_0_0_5_35#register.CRV)
* _Portable version:_ Change watchdog API to be able to update the window

Any suggestion? Have I missed anything?

**Comment by thedjnK** [comment]
You will have to work around this problem without the "hacky" version because as quoted on that documentation page, once the watchdog has been activated, those registers become read only: "The watchdog must be configured before it is started. After it is started, the watchdog’s configuration registers, which comprise registers CRV, RREN, and CONFIG, will be blocked for further configuration" and the only way to change them is to reset the CPU and configure the watchdog again.

**Comment by dleach02** [comment]
@carlescufi this seems like it should be converted to a discussion with the last comment basically being the answer. What do you think?
